### DISCUSSION TITLE
[QUESTION] How to copy file into a bound/attached PV

### DISCUSSION THREAD

**Discussion by ncecsc** [longhorn/longhorn#6781] [General]
## Question

Is there a way to copy a file (e.g. config.json) from the local system (/root) into a Longhorn PV with an attached statefulset pod without using kubectl cp?

I have seen https://github.com/longhorn/longhorn/issues/265 but mounting the volume in Linux only applies if the PV is available - which is not the case in my cluster.

A process in the statefulset container loads config.json into memory on boot, and writes it back when it stops. Before using Kubernetes, if we needed to update this file manually, we would stop the process, update the file, and restart the process. However due to how the container is set up (different mechanism than liveness checks), stopping the process forces kubernetes to restart the pod, which means we cannot use kubectl cp to copy the file into it.

Alternatively, is there a way to put kubernetes in some kind of maintenance mode where the pods keep running, but don't get restarted if a process in a container is killed?

## Environment

 - Longhorn version: 1.4.0
 - Kubernetes version: 1.26.3
 - Node config
   - OS type and version RHEL9
   - Kernel version 5.14.0-284.11.1.el9_2.x86_64
   - CPU per node: 32
   - Memory per node: 64 GB
   - Disk type Enterprise SSD
   - Network bandwidth and latency between the nodes: 2x 10Gb bonded via a 10 Gb switch
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal

**Comment by ChanYiLin** [comment]
I think in this case even if you use `kubectl cp` you still need to restart the process to load the config
1. Maybe you need some mechanism to watch the config file and reload it when it changes by forking another process.
2. Or build the image with the config file. In this case, you build a new image with CICD when you change the config file. Then you use rolling update to update the statefulset/deployment

Its more like the DevOps problem with Kubernetes, maybe you can find the best practice in this direction
I don't think Longhorn can help here.

**Comment by ncecsc** [comment]
> Its more like the DevOps problem with Kubernetes, maybe you can find the best practice in this direction
> I don't think Longhorn can help here.

Yeah, our developers changed the statefulset.yaml to include a debug hook which just launches a bash shell when doing kubectl apply, but still mounts the pvs. Then use kubectl cp, and finally comment out the debug hook and re-apply.
